llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp

```diff
@@ -497,6 +499,66 @@ static void allocateHSAUserSGPRs(CCState &CCInfo,
   // these from the dispatch pointer.
 }
 
+void AMDGPUCallLowering::lowerPreloadedParameter(
+    MachineIRBuilder &B, ArrayRef<Register> VRegs, Type *ArgTy,
+    uint64_t ArgOffset, Align Alignment,
+    ArrayRef<MCRegister> PreloadRegs) const {
+  MachineFunction &MF = B.getMF();
+  const GCNSubtarget *Subtarget = &MF.getSubtarget<GCNSubtarget>();
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const SIRegisterInfo *TRI = Subtarget->getRegisterInfo();
+  const DataLayout &DL = B.getDataLayout();
+
+  LLT ResTy = getLLTForType(*ArgTy, DL);
+  LLT ScalarTy = LLT::scalar(DL.getTypeSizeInBits(ArgTy));
```

Comment:
This isn't the most reliable way to get the register size, it should go through the calling convention type legalization. I guess this works out if we only allow preloaded on the most trivial types?

Reply:
Is there even a documented underlying calling convention? This also just mirrors what happens when we allocate in `allocatePreloadKernArgSGPRs`.

Reply:
There's a bunch of code to go from IR type to register type, which most calling convention code should be operating in terms of, not the original IR type (i.e. what's in CCValAssign and ISD::InputArg)